src/frequenz/sdk/util/asyncio.py

```diff
@@ -19,3 +20,22 @@ async def cancel_and_await(task: asyncio.Task) -> None:
         await task
     except asyncio.CancelledError:
         pass
+
+
+class NotSyncConstructible(AssertionError):
+    """Raised when object with async constructor is created in sync way."""
+
+
+class AsyncConstructible(ABC):
+    """Parent class for classes where part of the constructor is async."""
+
+    def __init__(self) -> None:
+        """Raise error when object is created in sync way.
+
+        Raises:
+            NotSyncConstructible: If this method is called.
+        """
+        raise NotSyncConstructible(
+            "This object shouldn't be created with default constructor. ",
+            "Check class documentation for more information.",
+        )
```

Comment:
maybe also add an abstract method, to standardize all uses?

``` python
@abstractmethod
async def async_new() -> Self:
  """Create an `AsyncConstructible` object."""
```

Reply:
I could do that, but the problem is that then number of constructor paramenters in async_new  method might different it he class that derives from it.

So in the `AsyncConstructible::async_new(cls: Type[T]) -> T `
But the classes that derives after it have different number of constructor arguments :confused:
And we get the pylint error:
```
Number of parameters was 1 in 'AsyncConstructible.async_new' and is now 4 in overridden 'StatusTracker.async_new' method
```
And the only way (I found) to avoid it is to create AsyncConstructibleConfig abstract class, and each class that derives from AsyncConstructible has to write its own config class that is passed to the constructor.

Reply:
As discussed it stays like this for now